clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java

```diff
@@ -231,6 +231,10 @@ SSLContext createSSLContext(SecurityStore keystore, SecurityStore truststore) th
             Password keyPassword = keystore.keyPassword != null ? keystore.keyPassword : keystore.password;
             kmf.init(ks, keyPassword.value().toCharArray());
             keyManagers = kmf.getKeyManagers();
+        } else if (kmfAlgorithm != null) {
```

Comment:
Could we not have this condition as a `||` in the previous `if` and make `keystore` initialization optional? Also, we need a test for this.